### PR TITLE
Fix problem with dash char in app name

### PR DESCRIPTION
**Pull request by davydovanton**
If you create new app with app architecture and name which include dash char (`-`) lotus generator created invalid names for modules. I.e this command:
```
$ lotus new test-app --architecture=app
```

Will create `Test-app` module:
``` ruby
module Test-app
  class Application < Lotus::Application
    # ...
  end
end
```

However the rails solved this problem. When you create `test-app` application rails generator set `TestApp` to all partials.

Therefore this commit change all `-` chars to `_` in `app_name` before calling `classify` method. However this changes increase allocation (because `tr` method call through `method_missing`). But I hope that it's
not critical issue on this part.

### CODE REVIEW

**Comment by davydovanton** [comment]
Also I think we need to check what will happen when you try to create new app with `test-app` name in container architecture. What do you think?

**Comment by joneslee85** [comment]
Yes please

For now I think it is okay as is but moving forward we should think dash is a separator for namespace, but till then

File: lib/lotus/commands/new/app.rb [hunk 1]
```diff
@@ -104,7 +104,7 @@ def upcase_app_name
         end
 
         def classified_app_name
-          Utils::String.new(app_name).classify
+          Utils::String.new(app_name).tr('-', '_').classify
```
Comment (AlfonsoUceda) [inline lib/lotus/commands/new/app.rb]: this should be fixed in lotus/utils repo

https://github.com/lotus/utils/blob/master/lib/lotus/utils/string.rb#L73
I think we have to add https://github.com/lotus/utils/blob/master/lib/lotus/utils/string.rb#L67

Comment (davydovanton) [inline lib/lotus/commands/new/app.rb]: Тo, I don't agree with you. If you open this link you see that `test-app` should be equal `Test::App`.
http://guides.rubygems.org/name-your-gem/

**Comment by davydovanton** [comment]
However, when I try to create `test-app` in container architecture I have similar problem. :disappointed:

**Comment by davydovanton** [comment]
@joneslee85 I fixed behavior for `lotus generate app test-app` command too. But I don't like this solution because I repeat code in `classified app_name` method. What do you think? May be is it better to replace this method to `Utils::String` or other place?

**Comment by davydovanton** [comment]
and also I have a strange [failed test](https://travis-ci.org/lotus/lotus/jobs/101341838#L256) for `Static` class.

**Comment by davydovanton** [comment]
@joneslee85 can I do anything to speed up merge of this PR? :)

**Comment by AlfonsoUceda** [comment]
@davydovanton one thing about this, when you do `lotus new test-app --architecture=app` do you expect `Test::App` or `TestApp`

**Comment by davydovanton** [comment]
I expect `TestApp` but in `app` architecture I will get `Web` :)

**Comment by AlfonsoUceda** [comment]
@davydovanton then this fix should be in lotus-utils, not here

**Comment by davydovanton** [comment]
I don't think so.
This changes fixed only `lotus new` command. `Utils::String.new('test-app').classify` should return `Test::App` but this behaviourus will breake lotus app. Because you can't create something like this:
``` ruby
module Test::App
  class Application < Lotus::Application
    # ...
  end
end
```

**Comment by AlfonsoUceda** [comment]
@davydovanton nop, that doesn't work like that.
```ruby
Lotus::Utils::String.new("Lotus::Utils").classify.to_s
 => "Lotus::Utils" #GOOD
Lotus::Utils::String.new("test-app").classify.to_s
 => "Test-app" # BAD
Lotus::Utils::String.new("test-app").classify.to_s
 => "TestApp" # EXPECTED
Lotus::Utils::String.new("test_app").classify.to_s
 => "TestApp" #GOOD
```

**Comment by davydovanton** [comment]
@AlfonsoUceda yes, you're right. But if you open this page - http://guides.rubygems.org/name-your-gem/ you can see that `'test-app'` string should be equals `Test::App`.

File: lib/lotus/commands/new/app.rb (hunk 1 above)
Comment (jodosha) [inline lib/lotus/commands/new/app.rb]: @davydovanton, @AlfonsoUceda is right `#classify` doesn't work properly and needs to be fixed.

```ruby
➜ bundle console
Resolving dependencies...
irb(main):001:0> require 'lotus/utils/string'
=> true
irb(main):002:0> Lotus::Utils::String.new('app-name').classify.to_s
=> "App-name"
```

It returns `"App-name"`, which is clearly an invalid Ruby class name. We expect `"App::Name"` as result.

**Comment by jodosha** [comment]
@davydovanton @AlfonsoUceda Unfortunately Lotus doesn't support deep naming. We have a convention that the top level module must have a depth of one. So `test-app` for now must be translated as `TestApp`, otherwise the world falls apart.

---

I don't like this convention, but we're forced to keep it because of `Configuration.for`:

  * https://github.com/lotus/controller/blob/master/lib/lotus/controller/configuration.rb#L90
  * https://github.com/lotus/view/blob/master/lib/lotus/view/configuration.rb#L90
  * https://github.com/lotus/assets/blob/master/lib/lotus/assets/configuration.rb#L74

This method is invoked when we _"duplicate"_ the frameworks (eg. from `Lotus::Action` we get `Web::Action`. With that method we can understand that an action that includes `Admin::Action` belongs to the `Admin` application.

---

Probably with 2.0 we'll remove this convention, but for now it has to stay. It isn't a good time to make radical changes like this.

**Comment by davydovanton** [comment]
@jodosha may be is better to replace `self.for` to utils module? If you agree, I can create PR.

and what about this PR? is this changes good for you?

**Comment by jodosha** [comment]
@davydovanton No, thanks but `Configuration.for` should stay as it is. That should be revisited later.

The `String#classify` fix works well :+1: - I'm gonna merge it now.

---

**We still need to respect `"app-name" => "AppName"` convention**. That means we should apply a transformation to the output of `#classify`: `Utils::String.new('app-name').gsub('::', '')`.

@davydovanton Can you please amend this PR? TY!

**Comment by davydovanton** [comment]
@jodosha done.

**Comment by davydovanton** [comment]
I have one failed build in travis (https://travis-ci.org/lotus/lotus/jobs/102324284). But this build failed because of tests cann't find `Assets` class.

**Comment by jodosha** [comment]
@davydovanton Thank you! :+1: I've restarted that failing CI Job. Waiting for it.

**Comment by jodosha** [comment]
@davydovanton Where are the tests for this? Can you please point to me where they are located? Thank you! :+1:

**Comment by davydovanton** [comment]
@jodosha sorry, I forgot this :disappointed: 
done :smiley:

**Comment by davydovanton** [comment]
@jodosha I don't know why but some builds failed with this error:
```
  1) Error:
Lotus::Commands::Generate::App::with valid arguments#test_0003_returns valid classified app name:
ArgumentError: App generator is only available for container architecture.
    /home/travis/build/lotus/lotus/lib/lotus/commands/generate/app.rb:110:in `assert_architecture!'
```
Do you have any ideas?

File: test/commands/generate/app_test.rb [hunk 2]
```diff
@@ -45,6 +45,11 @@
       end
     end
 
+    it 'returns valid classified app name' do
+      command = Lotus::Commands::Generate::App.new({}, 'test-app')
```
Comment (jodosha) [inline test/commands/generate/app_test.rb]: @davydovanton Can we use `"awesome-test-app"` instead?

I want to make sure that multiple dashes are handled well in the name. One day we could change `String#tr` with `String#sub` and the test still pass with `"test-app"`, but the feature would be broken.

File: lib/lotus/commands/new/app.rb [hunk 3]
```diff
@@ -104,7 +104,7 @@ def upcase_app_name
         end
 
         def classified_app_name
-          Utils::String.new(app_name).classify
+          Utils::String.new(app_name).classify.tr('::', '')
```
Comment (jodosha) [inline lib/lotus/commands/new/app.rb]: @davydovanton Is there any way to DRY this with `Commands::Generate::App#classified_app_name`?

Comment (davydovanton) [inline lib/lotus/commands/new/app.rb]: I thought about this. But we need to replace this method to utils repository for escape DRY. Also I thought about `Lotus::Generators::Generatable` but it incorrect place for this method too.

File: test/commands/generate/app_test.rb (hunk 2 above)
Comment (davydovanton) [inline test/commands/generate/app_test.rb]: good idea :+1:

File: lib/lotus/commands/new/app.rb (hunk 3 above)
Comment (jodosha) [inline lib/lotus/commands/new/app.rb]: @davydovanton Fine with that for now.

**Comment by jodosha** [comment]
@davydovanton This looks good now, merging :+1:

**Comment by davydovanton** [comment]
@joneslee85, @AlfonsoUceda, @jodosha thanks for help :+1:

**Pull request merged**